src/Oscoin/Time/Chrono.hs

```diff
@@ -0,0 +1,66 @@
+{-- | Little helpers geared towards making obvious in the types
+      how blocks are ordered.
+
+      The main idea here is /not/ to enforce some kind of ordering within
+      the wrapped functor 'f', but rather to state a /contract/ in the API
+      regarding which is the ordering of the elements.
+
+      The typical example where these two newtypes are useful is when
+      manipulating blocks: the 'BlockChain' data structure stores them
+      newest-first (i.e. the `tip` is at the head) but sometimes is useful
+      to have them in reverse order (for example when storing them in the
+      `BlockStore`, where it's important to store them starting by the oldest
+      first).
+--}
+
+module Oscoin.Time.Chrono
+    ( NewestFirst(..)
+    , OldestFirst(..)
+
+    , reverse
+    ) where
+
+import qualified Data.List as List
+import qualified Data.List.NonEmpty as NonEmpty
+import           Oscoin.Prelude hiding (reverse)
+
+
+-- | A simple typeclass which can \"talk\" about data structures which are
+-- \"chronologically ordered\", and as such can be reversed. We offer implementations
+-- for both `NewestFirst` and `OldestFirst` parametrised over `[]` and `NonEmpty`.
+-- This allows for quickly reversing the order while preserving a structure:
+--
+-- >>> toNewestFirst . reverse . OF $ [1,2,3]
+-- [3,2,1]
+--
+class ChronogicallyOrdered a where
+    type ReverseOf a :: *
+    reverse :: a -> ReverseOf a
+
+-- | A wrapper over a 'Foldable' functor 'f', typically a list or a non-empty
+-- one, where the API contract stats that elements are ordered from the newest
+-- to the oldest.
+newtype NewestFirst f a = NF { toNewestFirst :: f a }
+  deriving (Show, Eq, Functor, Foldable, Semigroup, Monoid)
+
+-- | A wrapper over a 'Foldable' functor 'f', typically a list or a non-empty
+-- one, where the API contract stats that elements are ordered from the oldest
+-- to the newest.
+newtype OldestFirst f a = OF { toOldestFirst :: f a }
+  deriving (Show, Eq, Functor, Foldable, Semigroup, Monoid)
```

Comment:
I think the correct terms are simply `Chronological` (OldestFirst) and `ReverseChronological` (NewestFirst)

Reply:
I am on the fence here: from one side I agree that the original naming is not the most elegant, but at the same time is brutally efficient: it states exactly how the data is ordered, and "oldest" vs "newest" helps building the intuition immediately. 

Conversely, `Chronological` and `ReverseChronological` don't really help in this sense (but I might be just me and the tunnel vision I have developed on these names).

Do you think the sweet spot could be to name the types the way you want but to keep the getters the same (i.e. retain `toOldestFirst` and `toNewestFirst`)?

If you have very strong sentiments here, I can surely change everything, that's not a problem 😉

Reply:
I think it's fine as it is : ) - the part that feels icky to me is that old/new sounds like we're talking about something which ages :derelict_house: . For blocks, recency or height might have been more appropriate.  I can live with it :wink: